src/Plugin/RulesExpression/RulesLoop.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\Plugin\RulesExpression\RulesLoop.
+ */
+
+namespace Drupal\rules\Plugin\RulesExpression;
+
+use Drupal\Core\TypedData\ListDataDefinitionInterface;
+use Drupal\rules\Engine\ActionExpressionContainer;
+use Drupal\rules\Engine\ExecutionMetadataStateInterface;
+use Drupal\rules\Engine\ExecutionStateInterface;
+use Drupal\rules\Engine\IntegrityViolationList;
+
+/**
+ * Holds a set of actions that are executed over the iteration of a list.
+ *
+ * @RulesExpression(
+ *   id = "rules_loop",
+ *   label = @Translation("Action set"),
+ *   form_class = "\Drupal\rules\Form\Expression\ActionSetForm"
+ * )
+ */
+class RulesLoop extends ActionExpressionContainer {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function executeWithState(ExecutionStateInterface $state) {
+    $list_data = $state->getVariable($this->configuration['list']);
```

Comment:
The list needs to be configurable as any other context such that users can use the data selector here and property paths are supported. Would it make sense to make the loop plugin context aware so we can our context mapping and form traits for the context mapping?